public/app/features/alerting/unified/rule-list/filter/SavedSearches.tsx

```diff
@@ -0,0 +1,956 @@
+
+    return defaultSearch ? [defaultSearch, ...others] : others;
+  }, [savedSearches]);
+
+  // Reset internal state when dropdown closes
+  useEffect(() => {
+    if (!isOpen) {
+      setActiveAction('idle');
+    }
+  }, [isOpen]);
+
+  // Focus "Save current search" button when dropdown opens
+  useEffect(() => {
+    if (isOpen && activeAction === 'idle' && saveButtonRef.current) {
+      // Small delay to ensure the dropdown is rendered
+      const timer = setTimeout(() => {
+        saveButtonRef.current?.focus();
+      }, 50);
+      return () => clearTimeout(timer);
+    }
+    return undefined;
+  }, [isOpen, activeAction]);
+
+  // Handle click outside to close dropdown
+  useEffect(() => {
+    const handleClickOutside = (event: MouseEvent) => {
+      if (
+        isOpen &&
+        dropdownRef.current &&
+        event.target instanceof Node &&
+        !dropdownRef.current.contains(event.target)
+      ) {
+        // Check if click is on a portal element (menu dropdown)
+        if (event.target instanceof Element) {
+          const isPortalClick =
+            event.target.closest('[data-popper-placement]') || event.target.closest('[role="menu"]');
+
+          if (!isPortalClick) {
+            setIsOpen(false);
+          }
+        } else {
+          setIsOpen(false);
+        }
+      }
+    };
+
+    if (isOpen) {
+      document.addEventListener('mousedown', handleClickOutside);
+    }
+
+    return () => {
+      document.removeEventListener('mousedown', handleClickOutside);
+    };
+  }, [isOpen]);
+
+  // Handle Escape key to close dropdown
+  useEffect(() => {
+    const handleKeyDown = (event: KeyboardEvent) => {
+      if (event.key === 'Escape' && isOpen) {
+        // If an action is active, cancel it first
+        if (activeAction !== 'idle') {
+          setActiveAction('idle');
+        } else {
+          setIsOpen(false);
+        }
+      }
+    };
+
+    document.addEventListener('keydown', handleKeyDown);
+    return () => document.removeEventListener('keydown', handleKeyDown);
+  }, [isOpen, activeAction]);
+
+  // Handlers
+  const handleToggle = useCallback(() => {
+    setIsOpen((prev) => !prev);
+  }, []);
+
+  const handleStartSave = useCallback(() => {
+    setActiveAction({ type: 'saving' });
+  }, []);
+
+  const handleCancelSave = useCallback(() => {
+    setActiveAction('idle');
+  }, []);
+
+  const handleSaveComplete = useCallback(
+    async (name: string): Promise<ValidationError | void> => {
+      const result = await onSave(name, currentSearchQuery);
+      if (!result) {
+        setActiveAction('idle');
+      }
+      return result;
+    },
+    [onSave, currentSearchQuery]
+  );
+
+  const handleStartRename = useCallback((id: string) => {
+    setActiveAction({ type: 'renaming', id });
+  }, []);
+
+  const handleCancelRename = useCallback(() => {
+    setActiveAction('idle');
+  }, []);
+
```

Comment:
If we want to have so detailed state tracking, it could be worth considering adding a reducer to handle the state management